tests/unit/s2n_protocol_version_getter_test.c

```diff
@@ -0,0 +1,395 @@
+/*
+* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include "api/s2n.h"
+#include "s2n_test.h"
+#include "testlib/s2n_testlib.h"
+#include "tls/s2n_tls.h"
+
+struct s2n_override_extension_ctx {
+    struct s2n_blob extension_blob;
+    int invoked_count;
+};
+
+static int s2n_override_supported_versions_cb(struct s2n_connection *conn, void *ctx)
+{
+    EXPECT_NOT_NULL(conn);
+    EXPECT_NOT_NULL(ctx);
+
+    struct s2n_override_extension_ctx *context = (struct s2n_override_extension_ctx *) ctx;
+    context->invoked_count += 1;
+
+    struct s2n_client_hello *client_hello = s2n_connection_get_client_hello(conn);
+    EXPECT_NOT_NULL(client_hello);
+
+    s2n_extension_type_id supported_versions_id = 0;
+    EXPECT_SUCCESS(s2n_extension_supported_iana_value_to_id(S2N_EXTENSION_SUPPORTED_VERSIONS, &supported_versions_id));
+
+    s2n_parsed_extension *supported_versions_extension = &client_hello->extensions.parsed_extensions[supported_versions_id];
+    supported_versions_extension->extension_type = S2N_EXTENSION_SUPPORTED_VERSIONS;
+    supported_versions_extension->extension = context->extension_blob;
+
+    return S2N_SUCCESS;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    DEFER_CLEANUP(struct s2n_cert_chain_and_key *chain_and_key = NULL, s2n_cert_chain_and_key_ptr_free);
+    EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key,
+            S2N_DEFAULT_TEST_CERT_CHAIN, S2N_DEFAULT_TEST_PRIVATE_KEY));
+
+    /* Safety */
+    {
+        EXPECT_FAILURE_WITH_ERRNO(s2n_connection_get_client_protocol_version(NULL), S2N_ERR_NULL);
+        EXPECT_FAILURE_WITH_ERRNO(s2n_connection_get_client_hello_version(NULL), S2N_ERR_NULL);
+        EXPECT_FAILURE_WITH_ERRNO(s2n_connection_get_server_protocol_version(NULL), S2N_ERR_NULL);
+        EXPECT_FAILURE_WITH_ERRNO(s2n_connection_get_actual_protocol_version(NULL), S2N_ERR_NULL);
+    }
+
+    /* Test protocol version getters on the server when a supported versions extension is received */
+    for (uint8_t server_version = S2N_TLS12; server_version <= S2N_TLS13; server_version++) {
+        for (uint8_t client_hello_version = S2N_SSLv3; client_hello_version <= S2N_TLS13; client_hello_version++) {
```

Comment:
Does the client hello version actually go all the way up to TLS1.3? I thought it was frozen at TLS1.2?

Reply:
No you're right, it's supposed to be frozen at TLS 1.2. However a client could still send a client hello with a version of TLS 1.3, so I thought it might be worth including that in the test.